core-v1/src/main/scala/data/Justified.scala

```diff
@@ -2,34 +2,126 @@ package com.rallyhealth.vapors.v1
+
+  trait Visitor[G[+_]] {
+
+    def visitConstant[V](justified: ByConst[V]): G[V]
+
+    def visitConfig[V](justified: ByConfig[V]): G[V]
+
+    def visitFact[V](justified: ByFact[V]): G[V]
+
+    def visitInference[V](justified: ByInference[V]): G[V]
   }
 
-  final case class ByConfig[+V](
+  def byConst[V](value: V): Justified[V] = ByConst(value)
+
+  final case class ByConst[V](value: V) extends Justified[V] {
+    override def visit[G[+_]](v: Visitor[G]): G[V] = v.visitConstant(this)
+    override def evidence: NoEvidence.type = NoEvidence
+    override val reason: String = s"[const: $value]"
+    override def configs: Seq[(String, Option[String])] = Seq.empty
+    override def productPrefix: String = "Justified.ByConst"
+  }
+
+  def byConfig[V](
+    value: V,
+    configKey: String,
+    configDescription: Option[String] = None,
+  ): Justified[V] = ByConfig(value, configKey, configDescription)
+
+  final case class ByConfig[V](
     value: V,
     configKey: String,
     configDescription: Option[String] = None,
   ) extends Justified[V] {
-    override def withValue[U](value: U): Justified[U] = copy(value = value)
+    override def evidence: NoEvidence.type = NoEvidence
+    override def visit[G[+_]](v: Visitor[G]): G[V] = v.visitConfig(this)
+    override val reason: String = s"[config: $configKey=$value]"
+    override def configs: Seq[(String, Option[String])] = Seq(configKey -> configDescription)
+    override def productPrefix: String = "Justified.ByConfig"
   }
 
-  final case class ByEvidence[+V](
+  def byFact[V](fact: TypedFact[V]): Justified[V] = ByFact(fact)
+
+  // TODO: Allow a lens for map operations?
+  final case class ByFact[V](fact: TypedFact[V]) extends Justified[V] {
+    override val evidence: SomeEvidence = SomeEvidence(NonEmptySet.of(fact))
+    override def value: V = fact.value
+    override val reason: String = s"[fact: ${fact.typeInfo.name}=${fact.value}]"
+    override def visit[G[+_]](v: Visitor[G]): G[V] = v.visitFact(this)
+    override def configs: Seq[(String, Option[String])] = Seq.empty
+    override def productPrefix: String = "Justified.ByFact"
+  }
+
+  def byInference[V](
+    reason: String,
     value: V,
-    evidence: Evidence,
+    sources: NonEmptyList[Justified[Any]],
+  ): Justified[V] = ByInference(reason, value, sources)
+
+  final case class ByInference[V](
+    reason: String,
+    value: V,
+    sources: NonEmptyList[Justified[Any]], // TODO: Should this be a NonEmptySet? How to sort it?
   ) extends Justified[V] {
-    override def withValue[U](value: U): Justified[U] = copy(value = value)
-    override def evidenceOrNone: Evidence = evidence
+    override def visit[G[+_]](v: Visitor[G]): G[V] = v.visitInference(this)
+    override lazy val configs: Seq[(String, Option[String])] = sources.map(_.configs).reduce // TODO: should this be sorted?
+    override lazy val evidence: Evidence = sources.foldLeft(Evidence.none)(_ | _.evidence) // TODO: Is this even valid?
```

Comment:
These seem like important TODOs. We aren't using justification yet, and I don't really have time to experiment with these, so I'm gonna leave them for now.